python/cuda_parallel/cuda/parallel/experimental/_utils/cai.py

```diff
@@ -7,10 +7,56 @@
 Utilities for extracting information from `__cuda_array_interface__`.
 """
 
+from typing import Optional, Tuple
 import numpy as np
 
 from ..typing import DeviceArrayLike
 
 
 def get_dtype(arr: DeviceArrayLike) -> np.dtype:
     return np.dtype(arr.__cuda_array_interface__["typestr"])
+
+
+def get_strides(arr: DeviceArrayLike) -> Optional[Tuple]:
+    return arr.__cuda_array_interface__["strides"]
+
+
+def get_shape(arr: DeviceArrayLike) -> Tuple:
+    return arr.__cuda_array_interface__["shape"]
+
+
+def is_contiguous(arr: DeviceArrayLike) -> bool:
+    shape, strides = get_shape(arr), get_strides(arr)
+
+    if strides is None:
+        return True
+
+    if sum(shape) == 0:
+        # array has no elements
+        return True
+
+    if all(dim == 1 for dim in shape):
+        # there is a single element:
+        return True
+
+    itemsize = get_dtype(arr).itemsize
+
+    if strides[-1] == itemsize:
```

Comment:
The numpy [docs](https://numpy.org/doc/2.1/reference/arrays.ndarray.html#internal-memory-layout-of-an-ndarray) explicitly say:

> It does not generally hold that self.strides[-1] == self.itemsize for C-style contiguous arrays or self.strides[0] == self.itemsize for Fortran-style contiguous arrays is true.

@leofang do you know when this can be the case (with the exception of size 0 and size 1 arrays)? And whether or not that applies to CUDA ndarrays?

Reply:
FYI CuPy has an internal routine to check C contiguity [here](https://github.com/cupy/cupy/blob/ed630b74d6eb74e73f3c2721e796bec568e233a0/cupy/_core/internal.pyx#L169-L188); for F contiguity the loop should be reverted. It is an alternative implementation of what's done here, including the size-0 and size-1 checks done above, and so I think what we have here should be fine. This check would ensure if there's any whole in the memory layout. If you like you can encapsulate the lines into a standalone helper function and call the helper twice (pass `shape` and `strides`, and then pass the reverted `shape[::-1]` and `strides[::-1]`).

Reply:
>  If you like you can encapsulate the lines into a standalone helper function

I tried this after you suggested it, and decided that the tiny bit of code repetition was better than the indirection introduced.